### DISCUSSION TITLE
What should we name the module build of PlayCanvas?

### DISCUSSION THREAD

**Discussion by willeastcott** [playcanvas/engine#2661] [General]
The Rollup build process now outputs `playcanvas.js` (ES5 UMD library) and `playcanvas.mjs` (ES6 module).

I just tried to convert a PlayCanvas engine example to use the module instead of the ES5 version. The example failed to load saying the `playcanvas.mjs` module was being served with MIME type "". So a default Apache install doesn't specify a MIME type for `mjs` files basically. Which kind of sucks, because `.mjs` seems fairly well accepted as the extension for JS modules.

To fix it, I had to edit my `httpd.conf` file to have:

```
    AddType text/javascript .mjs
```

But I think it's probably unrealistic for developers to have to go and edit their web server configs to get started with PlayCanvas.

We could have:

* playcanvas.js (ES6 module)
* playcanvas.es5.js (ES5 UMD library)

Or:

* playcanvas.js (ES5 UMD library)
* playcanvas.module.js (ES6 module)

I kind of like at least one build (ideally the recommended/best one being named just `playcanvas.js`. But, yeah, we could also be explicit everywhere and have:

* playcanvas.es5.js (ES5 UMD library)
* playcanvas.module.js (ES6 module)

Or even:

* playcanvas.es5.js (ES5 UMD library)
* playcanvas.es6.js (ES6 module)

Does anybody else have any thoughts/ideas/preferences about this?

**Comment by Maksims** [comment]
I've noticed as well lack of popularity on mjs file extension.

If js file is loaded not as module but with export, it will complain. So it is sensible to have `module` in filename.

There is basically two paths we can go:
1. By default encourage to go compatible way: serve ES5 as playcanvas.js. And ES6 version alongside.
2. By default serve new version to the community, with ES5 alongside if engine users target audience includes older platforms.

What is done behind the scene in Editor and publish is perhaps less relevant here.

I'm personally more inclined to modern version by default, and provide mechanism to properly notify if running on older platform. With good developer docs on this topic, for ES5 engine usage and hybrid solution.

**Reply by GamesTocaBoca** [reply]
I've noticed that the `.mjs` file extension isn't widely used, and loading a `.js` file without marking it as a module can cause export issues. To avoid this, it’s better to add module indicators in filenames. We have two options: either encourage a compatible approach by serving ES5 as `playcanvas.js` with an ES6 version alongside, or make the modern ES6 version the default with ES5 as a fallback. I lean towards the modern version by default, supported by strong developer documentation. Better to add a clear notification system for older platforms and ensure developers are well-informed.

**Comment by kungfooman** [comment]
Since `playcanvas.js` got established as ES5 for years now (with lots of examples all over the web), I would rather add the `es6` tag.

It is also quite easy to remember... "just put on `.es6.` at the end"

```
// ES5
https://code.playcanvas.com/playcanvas-latest.js
https://code.playcanvas.com/playcanvas-stable.js
https://code.playcanvas.com/playcanvas-latest.min.js
https://code.playcanvas.com/playcanvas-stable.min.js
// ES6
https://code.playcanvas.com/playcanvas-latest.es6.js
https://code.playcanvas.com/playcanvas-stable.es6.js
https://code.playcanvas.com/playcanvas-latest.min.es6.js
https://code.playcanvas.com/playcanvas-stable.min.es6.js
```

And then maybe a little example in `README.md`:

```js
    <script type="module">
        import * as pc from "https://code.playcanvas.com/playcanvas-stable.min.es6.js";
        window.pc = pc;
        // create a PlayCanvas application
        const canvas = document.getElementById('application');
        const app = new pc.Application(canvas);

        // ...
```
